src/http_proxy/mod.rs

```diff
@@ -0,0 +1,226 @@
+                }
+                Err(e) => {
+                    dbg!(e);
+                }
+            }
+        } else {
+            unimplemented!();
+        }
+    }
+}
+
+async fn tls_mitm_wrapper<T>(
+    client_stream: Framed<TcpStream, HttpClient>,
+    opening_request: Request<Vec<u8>>,
+    mitm: Arc<T>,
+) where
+    T: MitmLayer,
+{
+    tls_mitm(client_stream, opening_request, &CERT_AUTH, &KEY, mitm)
+        .await
+        .unwrap();
+}
+
+async fn tls_mitm<T>(
+    mut client_stream: Framed<TcpStream, HttpClient>,
+    opening_request: Request<Vec<u8>>,
+    cert_auth: &CA,
+    private_key: &PKey<Private>,
+    mitm: Arc<T>,
+) -> SafeResult
+where
+    T: MitmLayer
+{
+    let (host, port) = target_host_port(&opening_request);
+    let (mut target_stream, server_certificate) = connect_to_target(&host, &port).await;
+    client_stream
+        .send(
+            &Response::builder()
+                .status(200)
+                .version(http::Version::HTTP_11)
+                .body(Vec::new())
+                .unwrap(),
+        )
+        .await?;
+
+    let certificate = spoof_certificate(&server_certificate, cert_auth).unwrap();
+    let identity = native_identity(&certificate, private_key);
+    let mut client_stream = convert_to_tls(client_stream, identity).await;
+    let proxy_connection: HeaderName = HeaderName::from_lowercase(b"proxy-connection").unwrap();
+
+    while let Some(request) = client_stream.next().await {
+        let mut request = request.unwrap();
+        match mitm.capture_request(&request) {
+            RequestCapture::CircumventedResponse(response) => {
+                client_stream.send(&response).await.unwrap();
+                continue;
+            }
+            RequestCapture::ModifiedRequest(new_request) => request = new_request,
+            RequestCapture::Continue => {}
+        }
+
+        *request.uri_mut() = request.uri().path().parse().unwrap();
+        request.headers_mut().remove(&proxy_connection);
+        target_stream.send(&request).await?;
+
+        let mut response = target_stream.next().await.unwrap()?;
+        match mitm.capture_response(&request, &response) {
+            ResponseCapture::ModifiedResponse(new_response) => {
+                response = new_response;
+            }
+            ResponseCapture::Continue => {}
+        }
+        client_stream.send(&response).await.unwrap();
+    }
+
+    Ok(())
+}
+
+async fn convert_to_tls(
+    client_stream: Framed<TcpStream, HttpClient>,
+    identity: native_tls::Identity,
+) -> Framed<TlsStream<TcpStream>, HttpClient> {
+    let client_stream = client_stream.into_inner();
+    let client = TlsAcceptor::from(native_tls::TlsAcceptor::new(identity).unwrap());
+    let client_stream = client.accept(client_stream).await.unwrap();
+    Framed::new(client_stream, HttpClient)
+}
+
+fn target_host_port(request: &Request<Vec<u8>>) -> (String, String) {
+    let host_header = String::from_utf8(Vec::from(
```

Comment:
You can save an allocation here. Instead of allocating a new vector on the heap you can slice directly into the `&[u8]` that `as_bytes` returns. You call `to_string()` at the end anyway.

Reply:
Forgot to mention that you can use `str::from_utf8(&[u8])`